doc/policies.md

```diff
@@ -0,0 +1,180 @@
+# APIcast policies
+
+The behaviour of APIcast is customizable via policies. A policy basically tells
+APIcast what it should do in each of the nginx phases. This document details
+how policies work and how you can write your own and integrate them into
+APIcast.
+
+- [**Policies**](#policies)
+- [**Policy chains**](#policy-chains)
+- [**APIcast default policies**](#apicast-default-policies)
+- [**Write your own policy**](#write-your-own-policy)
+- [**Integrate your policies**](#integrate-your-policies)
+
+
+## Policies
+
+A policy tells APIcast what it should do in each of the nginx phases: `init`,
+`init_worker`, `rewrite`, `access`, `balancer`, `header_filter`, `body_filter`,
+`post_action`, and `log`.
+
+Policies can share data between them. They do that through what we call the
+`context`. Policies can read from and modify that context in every phase.
+
+
+## Policy chains
+
+Policies can be combined using a policy chain. A policy chain is simply a
+sorted list of policies.
+
+The way policy chains work is as follows: suppose that we have a policy A that
+describes what to do in the `rewrite` and `header_filter` phases and a policy B
+that describes what to run in `access` and `header_filter`. Assume also that
+when describing the chain, we indicate that policy A should be run before
+policy B. When APIcast receives a request, it will check the policy chain
+described to see what it should run on each phase:
+- rewrite: execute what policy A specifies for this phase.
+- access: execute what policy B specifies for this phase.
+- balancer: do nothing. Neither policy A nor B describe what to do.
+- header_filter: execute first what policy A specified for this phase and then
+  what B described. Remember that policy chains define an order. And we
+  specified that policy A comes before policy B.
+- body_filter: do nothing. Neither policy A nor B describe what to do.
+- post_action: do nothing. Neither policy A nor B describe what to do.
+- log: do nothing. Neither policy A nor B describe what to do.
+
+Notice that we did not indicate what APIcast does in the `init` and the
+`init_worker` phases. The reason is that those two are not executed in every
+request. `init` is executed when APIcast boots, and `init_worker` when each
+of each of its workers start.
+
+### Types
+
+There are two types of policy chains in APIcast: per-service chains and a
+global chain. Both of them can be configured.
+
+As the name indicates, per-service chains allow us to define a specific chain
+for a given service. This means that we can apply different policies to our
+services, or the same policies but configured differently, or in a different
+order. On the other hand, there's only one global chain, and the behavior it
+defines applies to all the services.
```

Comment:
so, policies defined by both chains will be run?
How is order defined between policies in the global chain and policies in the service chain....

Reply:
Yes, per-service are included as part of the global one.

Reply:
The order of policies in the global chain is configurable. The order of policies in per-service chains is also configurable.